dhcp-relay/dhcp_kern_xdp.c

```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 
 
 	} else if (dhcp->op == DHCP_REPLY && (eth->h_dest[0] != 0xff
-		|| eth->h_dest[1] != 0xff
-		|| eth->h_dest[2] != 0xff
-		|| eth->h_dest[3] != 0xff
-		|| eth->h_dest[4] != 0xff
-		|| eth->h_dest[5] != 0xff)) {
+			|| eth->h_dest[1] != 0xff
+			|| eth->h_dest[2] != 0xff
+			|| eth->h_dest[3] != 0xff
+			|| eth->h_dest[4] != 0xff
+			|| eth->h_dest[5] != 0xff)) {
 
 		/* Response from server received as unicast */
 
 		bpf_printk("Unicast packet received, opcode %i, hops %i", dhcp->op, dhcp->hops);
 
-		/* FIXME: Add code for reply packets
-		 * Basically:
-		 * - Set dest and src MAC
-		 * - Add VLAN tags
-		 * - Remove option 82
-		 * - Use XDP_TX (or XDP_REDIRECT) to send the response
-		 * to the end user
-		 */
-
 		struct collect_vlans *new_vlans;
 		new_vlans = bpf_map_lookup_elem(&client_vlans, &client_mac);
 		if (new_vlans == NULL) {
 			bpf_printk("Could not find map entry for MAC %i", client_mac);
 			goto out;
 		}
 
-		bpf_printk("Found map entry for MAC %i", client_mac);
+		bpf_printk("Found map entry for MAC %x", client_mac);
 
-	}
+		/* Set destination MAC */
+		memcpy(eth->h_dest, dhcp->chaddr, ETH_ALEN);
 
-	/* Check hops */
-	if (dhcp->hops > 16) {
-		bpf_printk("Max hops exceeded, discarding packet");
-		rc = XDP_ABORTED;
-		goto out;
-	}
+		/* Set source MAC */
+		memcpy(eth->h_source, relay_hwaddr, ETH_ALEN);
 
-	/* Increment hops */
-	dhcp->hops++;
+		/* Set destination IP */
+		ip->daddr = IP_ADDR_BCAST;
+
+		/* Set source IP */
+		ip->saddr = *relay_agent_ip;
```

Comment:
Jumping back and forth between modifying L2 and L3 headers makes the code hard to follow. Also this function is way too long, it should be split: let's turn it into separate `handle_request()` and `handle_reply()` functions, and also split the VLAN tag insertion/removal into a separate helper.